### ISSUE TITLE
Istio provides read-only access across namespaces, even when using an authentication method with limited access

### ISSUE THREAD

**Issue by mlbiam** [kiali/kiali#7125]
### Describe the bug

When setting up Istio in a multi-tenant environment, I have access to all configurations available to isitio for read-only, even if the token used to authenticate to Istio has no access to that namespace.

When I originally wrote the impersonation/auth header code it was written as such as long as the user's auth was available, that's what was used to communicate with the API server.  It looks like at some point that changed to using the container's `ServiceAccount`, which has broad read-only cluster wide permissions.  Depending on what was changed, I figure there are one of two approaches to correct this:

1. Revert the code that uses the SA to instead use the user's credentials
2. Build in access controls into Kiali

option 1 is the better security solution, since it relies on the API server for security and doesn't require any kind of configuration.  Option 2 can be done with `SubjectAccessReview` requests to eliminate the need for adding a configuration.  

I'm happy to submit a PR on this.  I know bandwidth is very limited.  Is there a preferred method or guidance?  It's been a couple of years since I looked at the codebase.

### Expected Behavior

In a multi-tenant environment, users should not be able to read from other namespaces.  

### What are the steps to reproduce this bug?

1. Deploy Kubernetes & Istio with an authenticating reverse proxy
4. Deploy kiali behind the authenticating reverse proxy
5. login to kilali with a token that does not provide access to all namespaces

### Environment
*Learn about how to determine versions [here](https://kiali.io/docs/faq/general/#how-do-i-determine-what-version-i-am-running).*

- **Kiali version:** v1.80.0 (3a1118fa7fd23a155d19a250c11a4574bf0c094e)
- **Istio version:**     1.20.2 
- **Kubernetes impl:**      Civo/K3s
- **Kubernetes version:**      v1.28.2+k3s1 
- **Other notable environmental factors:** authenticating reverse proxy is openunison

**Comment by jmazzitelli** [comment]
> that's what was used to communicate with the API server. It looks like at some point that changed to using the container's ServiceAccount, which has broad read-only cluster wide permissions. Depending on what was changed, I figure there are one of two approaches to correct this:

This sounds like a result of the work @nrfox is doing for multi-cluster. Ping @nrfox

**Comment by mlbiam** [comment]
doing some digging this appears to be the logic that's causing the issue:

https://github.com/kiali/kiali/blob/82719968b06f7d69e5955efe08b6a3d59f1e8b8f/business/namespaces.go#L337

Kiali is falling back to the `ServiceAccount` when unauthorized.  This may make for a better user experience but opens an unintended security hole.  The two contexts (`userClients` and `kialiSAClients`) but it appears that, at least in this instance, a 403 results into a repeat operation with the `ServiceAccount`

**Comment by nrfox** [comment]
> When I originally wrote the impersonation/auth header code it was written as such as long as the user's auth was available, that's what was used to communicate with the API server. It looks like at some point that changed to using the container's ServiceAccount, which has broad read-only cluster wide permissions.

I'm not sure how far back this goes but for at least as long as I have known (2-3 years), reads are done from the local kube cache with the Kiali Service Account and access is gated by checking if the user can GET the namespace. Writes to the API server always use the user's token and not the Kiali Service Account. From what I can tell, when Kiali falls back to use the Kiali Service Account to get all namespaces, it should still filter the results by performing a `GET` request using the user's token for every namespace that got returned by the Kiali Service Account: https://github.com/kiali/kiali/blob/82719968b06f7d69e5955efe08b6a3d59f1e8b8f/business/namespaces.go#L625-L653

> Build in access controls into Kiali

Kiali is already doing this (for better or worse) by checking the user's access to the namespace before returning the results. My guess is that the kube cache was added to help with performance issues associated with simply making calls to the API server using the user's token and these namespace controls we're added then.

How exactly is Kiali configured in your setup? Are you using the `header` auth strategy or the `openid` auth strategy with an `api_proxy`?

**Comment by jmazzitelli** [comment]
This is documented in the Kiali docs: https://kiali.io/docs/configuration/rbac/#granting-access-to-namespaces

```
Granting access to namespaces

In general, Kiali will give read access to namespaces where the logged
in user is allowed to “GET” its definition – i.e. the user is allowed to do
a GET call to the api/v1/namespaces/{namespace-name} endpoint of
the cluster API. Users granted the LIST verb would get access to all
namespaces of the cluster (that’s a GET call to the api/v1/namespaces
endpoint of the cluster API).

You, probably, will want to have this small ClusterRole to help you in
authorizing individual namespaces in Kiali
...:
```

**Comment by nrfox** [comment]
> There needs to be a SubjectAccessReview on the objects being requested for this to be done correctly.

I agree the current model is pretty "coarse" and could be greatly improved by doing `SubjectAccessReview`s. Given the current limitations, we still need to make sure that there's not a bug in the `header` strategy. A user who can `GET` on `<yournamespace>` should be able to see all the applications in that namespace but a user who cannot `GET` on `<yournamespace>` shouldn't even see that namespace in the namespace selector dropdown much less see any application info for it. Is that what you're seeing?

**Comment by mlbiam** [comment]
> A user who can GET on <yournamespace> should be able to see all the applications in that namespace but a user who cannot GET on <yournamespace> shouldn't even see that namespace in the namespace selector dropdown much less see any application info for it. Is that what you're seeing?

Correct, see that I'm not able to retrieve the `openunison` namespace with an unprivileged user, but i am able to list all namespaces:

```
➜  helm-charts git:(master) k auth whoami 
ATTRIBUTE   VALUE
Username    00u1aukqiuRTDHwyM357
Groups      [users approvers-k8s-k8s-test-unpriv-ns k8s-namespace-administrators-k8s-test-unpriv-ns-internal system:authenticated]
➜  helm-charts git:(master) k get ns openunison
Error from server (Forbidden): namespaces "openunison" is forbidden: User "00u1aukqiuRTDHwyM357" cannot get resource "namespaces" in API group "" in the namespace "openunison"
➜  helm-charts git:(master) k get ns
NAME                   STATUS   AGE
blackhole              Active   17m
cert-manager           Active   11d
default                Active   11d
istio-app              Active   8d
istio-hello-world      Active   10d
istio-system           Active   11d
kube-node-lease        Active   11d
kube-public            Active   11d
kube-system            Active   11d
kubernetes-dashboard   Active   11d
local-path-storage     Active   19m
mariadb                Active   17m
openunison             Active   11d
test-unpriv-ns         Active   67s
➜  helm-charts git:(master) 
```

but if i'm logged into kiali with the header strategy the same user is able to lookup configurations inside of the `openunison` namespace:

<img width="1800" alt="Screenshot 2024-02-20 at 8 03 33 PM" src="https://github.com/kiali/kiali/assets/8249283/9d7442fc-6f73-4865-825c-d9fbd7ed7cfe">

**Comment by nrfox** [comment]
> Correct, see that I'm not able to retrieve the openunison namespace with an unprivileged user, but i am able to list all namespaces

Ah ok so your user has `LIST` permissions for namespaces but can't `GET` the namespace. So I was wrong in my comment above. It looks like the way this is implemented in Kiali, if you've got any read access at all to the namespace either with `LIST` or `GET` then you'll be able to see info from that namespace in Kiali. I think this is "working as intended" but could you test with an older version of Kiali to ensure that something hasn't changed recently? `v1.65` would be good.

We can also open a separate enhancement for stricter access controls. I thought we had one already but I can't find one.

**Comment by jmazzitelli** [comment]
> It looks like the way this is implemented in Kiali, if you've got any read access at all to the namespace either with `LIST` or `GET` then you'll be able to see info from that namespace in Kiali. I think this is "working as intended"

That's what I was trying to say with my comment above : https://github.com/kiali/kiali/issues/7125#issuecomment-1952929076

**Comment by mlbiam** [comment]
> but could you test with an older version of Kiali to ensure that something hasn't changed recently? v1.65 would be good.

It looks like the breaking change was part of the 1.75 release.  

1.75:

<img width="1800" alt="Screenshot 2024-02-21 at 1 48 18 PM" src="https://github.com/kiali/kiali/assets/8249283/326ab004-13db-46b9-b967-107ed8e1f4c5">


1.74:
<img width="1800" alt="Screenshot 2024-02-21 at 1 44 33 PM" src="https://github.com/kiali/kiali/assets/8249283/a0dec748-cfd7-4485-ad91-28b4d7067980">

@jmazzitelli sorry, wasn't ignoring your post.  just wanted to make sure we all had the same assumptions in place. 

> In general, Kiali will give read access to namespaces where the logged
in user is allowed to “GET” its definition – i.e. the user is allowed to do
a GET call to the api/v1/namespaces/{namespace-name} endpoint of
the cluster API. Users granted the LIST verb would get access to all
namespaces of the cluster (that’s a GET call to the api/v1/namespaces
endpoint of the cluster API).

So I'd argue this is an anti-pattern from an access management standpoint.  Being able to list namespaces is a requirement for nearly every other kubernetes dashboard and the workaround RBAC for the above statement would break that assumption.  i know that performance could be an issue, and would be comfortable saying the namespace is the boundary but there should be a `SubjectAccessReview` of one of the objects being looked up such as deployments or configmaps.  That could also help centralize the logic.

**Comment by mlbiam** [comment]
Went through the commits comparing the 1.74.1 tag <-> 1.75.0 and there are only change that appears to be the culprit (or really related at all) is:

https://github.com/kiali/kiali/commit/45d72534b0556c0425f52a1488811eb372c92e43

But this only happens when the cluster is deemed to be remote. Is there maybe a bug that clusters are always being treated as remote?

**Comment by mlbiam** [comment]
Tried backing out https://github.com/kiali/kiali/issues/7125#issuecomment-1957920303 locally and testing, same result as 1.75.0.

**Comment by nrfox** [comment]
> Went through the commits comparing the 1.74.1 tag <-> 1.75.0 and there are only change that appears to be the culprit (or really related at all) is:

Nothing really jumps out at me either. What version of Istio are you testing with 1.74.1? iirc 1.74 doesn't work with v1.20 so maybe that's the reason you aren't seeing applications? Are there other errors on the page? Given that _any_ read access _should_ allow you to view what is in that namespace in Kiali I'm surprised that 1.74 behaves differently.

> So I'd argue this is an anti-pattern from an access management standpoint. Being able to list namespaces is a requirement for nearly every other kubernetes dashboard and the workaround RBAC for the above statement would break that assumption.

There's another access knob in Kiali: [accessible_namespaces](https://kiali.io/docs/configuration/namespace-management/#accessible-namespaces) which limits the namespaces Kiali itself has access to and therefore what all users have access to through Kiali. Although you would still run into the same issue if a user could LIST all namespaces but not GET individual ones. It's just that using `accessible_namespaces`, what they see would be limited to what Kiali also has access to.

Personally I'm not opposed to changing this but I'd be wary of breaking existing users who might rely on this behavior. We could add yet another access related config option to Kiali to enable this. There's already a lot of access related config flags so I'm a little wary of doing that but again I'm not opposed. @jmazzitelli thoughts?

> there should be a SubjectAccessReview of one of the objects being looked up such as deployments or configmaps.

How do you choose what to lookup? If the namespace is the boundary then it could be confusing that you _also_ need some arbitrary access to something else inside that namespace. I'd much rather go whole hog and do a `SubjectAccessReview` for everything Kiali returns. Metrics/traces should be included in that as well by using the user's token to query both those datastores.

**Comment by jmazzitelli** [comment]
Question: Why does a user of Kiali need LIST permission for namespaces?

**Comment by mlbiam** [comment]
@nrfox - So i totally messed up my testing.  I was checking the Applications tab and I think you're right about just the wrong version.  I noticed that the configs tab was giving me different results so i'm going to test again tomorrow now that I'm home.

@jmazzitelli  - A Kiali user might not need LIST permissions, but other dashboards (such as the Kubernetes Dashboard and KubeApps) need LIST permissions for namespaces.  So removing that would break the other two dashboards.

**Comment by jshaughn** [comment]
Thanks @mlbiam , both Nick and Mazz are out for few days, I'm sure they will follow up when they can.

**Comment by jshaughn** [comment]
@mlbiam @jmazzitelli @nrfox : 

Having read through this it seems like Kiali is behaving as intended, so I converted the Bug label to Enhancement.  The Enhancement, I think, would be a new config option that would affect the handling of the LIST permission.  By default Kiali would operate as it does today, to maintain backward compatibility.  But optionally we could force GET permission to be required for read access to a namespace.  In this mode, without GET a user would not see the namespace in the namespace dropdown.

Does that reflect the need?

As mentioned above, Istio and Kiali already support, and recommend the use of Discovery Selectors to narrow the total set of accessible namespaces for the mesh.

**Comment by jshaughn** [comment]
@mlbiam Please read the above and let us know if this is still a need, thanks.

**Comment by mlbiam** [comment]
Sorry, this fell off the map, though i do need to revisit.  I disagree @jshaughn that the functionality is acceptable as is.  The problem with discovery selectors is they're static in the configuration which means you'd need a separate Kiali per tenant.  That would be un-manageable.  While openunison could be customized to correctly route to the correct Kiali at run time, it would be nearly impossible for most other reverse proxies built with static configurations.  This is on top of the resource requirements of having a different kiali instance for each tenant.

**Comment by jmazzitelli** [comment]
> you'd need a separate Kiali per tenant

Maybe we have different definitions of "tenant" but, Kiali has always been "one Kiali for each tenant in a multi-tenant cluster" -- meaning for each Istio service mesh within the cluster, there is one Kiali.

**Comment by mlbiam** [comment]
> meaning for each Istio service mesh within the cluster, there is one Kiali.

I'd argue that's a really narrow definition and is prone to security issues. It's also not how the original impersonation code was implemented. Every multi tenant kubernetes implementation I've done has worked on the model that the management systems are shared based on kubes security model.

**Comment by jmazzitelli** [comment]
Well, I can say without question that Kiali since day one has been designed (and implemented) to support/manage/observe one and only one Istio Service Mesh. It was never the intention for Kiali to be able to observe multiple service meshes (indeed, Kiali only ever talks to a single istiod, for example). Even the multi-cluster stuff that is currently being worked on only sees a single mesh (even when crossing cluster boundaries).

**Comment by mlbiam** [comment]
> Well, I can say without question that Kiali since day one has been designed (and implemented) to support/manage/observe one and only one Istio Service Mesh. 

I'm not saying anything different. One service mesh, multiple users. Multi tenant. One cluster where different teams own different namespaces and are only able to create and read kubernetes objects in their namespace.

**Comment by jmazzitelli** [comment]
> I'm not saying anything different. One service mesh, multiple users. Multi tenant. One cluster where different teams own different namespaces and are only able to create and read kubernetes objects in their namespace.

OK. That's why I asked earlier if maybe our definition of "tenant" is different. In "Kiali parlance" (as it is used particularly when working within a Maistra (aka OpenShift Service Mesh) environment), a "tenant" is a service mesh within the cluster (where namespace N1 and N2 are in service mesh 1, and namespace N3 and N4 are in service mesh 2 - so the one cluster has two service mesh "tenants"). I think we are just using the term "tenant" differently. Multi-tenancy in OSSM/Kiali therefore refers to different Istio service meshes dispersed within different namespaces within a single cluster.

**Comment by jshaughn** [comment]
> . I disagree @jshaughn that the functionality is acceptable as is. The problem with discovery selectors is they're static...

@mlbiam I didn't say that the behavior was acceptable as is, only that Kiali was behaving as currently designed.  And I didn't say Discovery Selectors solved the problem, I only mentioned, as an aside, that they exist as the recommended way to scope namespaces for the entire mesh, and that Kiali respects them.  What I did say is that maybe Kiali needed a new option to be more strict, optionally requiring 'get', in addition to 'list', for namespace access.  We could keep default behavior as-is as to not break existing users.  Or, we could deprecate the current behavior and require an option be set to maintain the current 'list' behavior.

**Comment by chinaran** [comment]
Any updates on this issue?

**Comment by jshaughn** [comment]
Hi @chinaran. It looks like we never really agreed on a path forward, so I think we still have the same behavior. I don't think Kiali will move in the direction of SubjectAccessReview, it seems too granular for what we want to do. What I would propose (again), is making Kiali configurable in a way that we either:
- Assume namespace GET if we have namespace LIST permission
  - this is what we do today if Kiali is running in cluster-wider mode
- Require namespace GET
  - This would be a more strict mode, such that a user would only have access to namespaces for which he has GET permission

Note that in both cases Kiali also needs access.

So, in this new strict-get mode, for a single Kiali managing a mesh with several namespaces, users could have LIST perm to all of the namespaces but only have GET to a single namespace. And Kiali would only show them, for example, the single namespace in the dropdown.

Would this be helpful to you?

**Comment by jshaughn** [comment]
I decided to investigate this proposal further with a Draft PR: https://github.com/kiali/kiali/pull/9558

**Comment by jshaughn** [comment]
Scheduled for Kiali 2.25 release.
